pulsar-broker/src/main/java/org/apache/pulsar/common/naming/FlowOrQpsEquallyDivideBundleSplitAlgorithm.java

```diff
@@ -0,0 +1,114 @@
+            }
+
+            double bundleThroughput = 0;
+            double bundleMsgRate = 0;
+            Map<Long, Double> hashAndMsgMap = new HashMap<>();
+            Map<Long, Double> hashAndThroughput = new HashMap<>();
+            Map<Long, String> hashAndTopic = new HashMap<>();
+            List<Long> topicNameHashList = new ArrayList<>(topics.size());
+            for (String topic : topics) {
+                TopicStatsImpl topicStats = topicStatsMap.get(topic);
+                if (topicStats == null) {
+                    continue;
+                }
+                double msgRateIn = topicStats.getMsgRateIn();
+                double msgRateOut = topicStats.getMsgRateOut();
+                double msgThroughputIn = topicStats.getMsgThroughputIn();
+                double msgThroughputOut = topicStats.getMsgThroughputOut();
+                double msgRate = msgRateIn + msgRateOut;
+                double throughput = msgThroughputIn + msgThroughputOut;
+
+                if (msgRate <= 0 && throughput <= 0) {
+                    // Skip empty topic
+                    continue;
+                }
+
+                Long hashCode = bundle.getNamespaceBundleFactory().getLongHashCode(topic);
+                topicNameHashList.add(hashCode);
+                hashAndMsgMap.put(hashCode, msgRate);
+                hashAndThroughput.put(hashCode, throughput);
+                hashAndTopic.put(hashCode, topic);
+                bundleThroughput += throughput;
+                bundleMsgRate += msgRate;
+            }
+
+            if (topicNameHashList.size() < 2
+                    || (bundleMsgRate < (loadBalancerNamespaceBundleMaxMsgRate * (1 + diffThreshold))
+                    && bundleThroughput < (loadBalancerNamespaceBundleMaxBandwidthBytes * (1 + diffThreshold)))) {
+                return CompletableFuture.completedFuture(null);
+            }
+
+            Collections.sort(topicNameHashList);
+            List<Long> splitResults = new ArrayList<>();
+            double bundleMsgRateTmp = 0;
+            double bundleThroughputTmp = 0;
+            for (int i = 0; i < topicNameHashList.size(); i++) {
+                long topicHashCode = topicNameHashList.get(i);
+                bundleThroughputTmp += hashAndThroughput.get(topicHashCode);
+                bundleMsgRateTmp += hashAndMsgMap.get(topicHashCode);
+
+                if (bundleMsgRateTmp > loadBalancerNamespaceBundleMaxMsgRate
+                        || bundleThroughputTmp > loadBalancerNamespaceBundleMaxBandwidthBytes) {
+                    long splitStart = i > 0 ? topicNameHashList.get(i - 1) : topicHashCode;
+                    long splitEnd = i > 0 ? topicHashCode : topicNameHashList.get(i + 1);
+                    long splitMiddle = splitStart + (splitEnd - splitStart) / 2;
+                    splitResults.add(splitMiddle);
+                    bundleMsgRateTmp =  hashAndMsgMap.get(topicHashCode);
```

Comment:
Seems not right if i == 0 here ? Please help point out the unit test covering this case.

Reply:
fixed

Reply:
@Jason918 PTAL,thanks!